File: server/internal/portfolio/db/reports.go
package portfoliodb

import (
	"context"
	"time"

	"tallyo/internal/database/dbgen"
	"tallyo/internal/database/dbutil"
	"tallyo/internal/portfolio"
)

func (s *Store) UpsertReport(ctx context.Context, report portfolio.AssetReport) error {
	fetchedAt := report.FetchedAt.UTC()
	if report.FetchedAt.IsZero() {
		fetchedAt = time.Now().UTC()
	}
	return s.q.UpsertAssetAnalysisReport(ctx, dbgen.UpsertAssetAnalysisReportParams{
		AssetID:                     report.AssetID,
		Category:                    report.Category,
		GroupName:                   report.Group,
		CashPosition:                report.CashPosition,
		StockPosition:               report.StockPosition,
		BondPosition:                report.BondPosition,
		PreferredPosition:           report.PreferredPosition,
		ConvertiblePosition:         report.ConvertiblePosition,
		OtherPosition:               report.OtherPosition,
		SectorRealEstate:            report.SectorRealEstate,
		SectorConsumerCyclical:      report.SectorConsumerCyclical,
		SectorBasicMaterials:        report.SectorBasicMaterials,
		SectorConsumerDefensive:     report.SectorConsumerDefensive,
		SectorTechnology:            report.SectorTechnology,
		SectorCommunicationServices: report.SectorCommunicationServices,
		SectorFinancialServices:     report.SectorFinancialServices,
		SectorUtilities:             report.SectorUtilities,
		SectorIndustrials:           report.SectorIndustrials,
		SectorEnergy:                report.SectorEnergy,
		SectorHealthcare:            report.SectorHealthcare,
		EquitySector:                report.EquitySector,
		FetchedAt:                   fetchedAt,
	})
}

func (s *Store) ReportsByAssetIDs(ctx context.Context, assetIDs []int64) (map[int64]portfolio.AssetReport, error) {
	if len(assetIDs) == 0 {
		return map[int64]portfolio.AssetReport{}, nil
	}
	rows, err := s.q.AssetAnalysisReportsByAssetIDs(ctx, dbgen.AssetAnalysisReportsByAssetIDsParams{AssetIds: assetIDs})
	toReportByAssetID := func(row dbgen.AssetAnalysisReportsByAssetIDsRow) (int64, portfolio.AssetReport) {
		report := reportFromRow(row.AssetAnalysisReport)
		return report.AssetID, report
	}
	return dbutil.AssociateRows(rows, err, toReportByAssetID)
}

func reportFromRow(row dbgen.AssetAnalysisReport) portfolio.AssetReport {
	return portfolio.AssetReport{
		AssetID:                     row.AssetID,
		Category:                    row.Category,
		Group:                       row.GroupName,
		CashPosition:                row.CashPosition,
		StockPosition:               row.StockPosition,
		BondPosition:                row.BondPosition,
		PreferredPosition:           row.PreferredPosition,
		ConvertiblePosition:         row.ConvertiblePosition,
		OtherPosition:               row.OtherPosition,
		SectorRealEstate:            row.SectorRealEstate,
		SectorConsumerCyclical:      row.SectorConsumerCyclical,
		SectorBasicMaterials:        row.SectorBasicMaterials,
		SectorConsumerDefensive:     row.SectorConsumerDefensive,
		SectorTechnology:            row.SectorTechnology,
		SectorCommunicationServices: row.SectorCommunicationServices,
		SectorFinancialServices:     row.SectorFinancialServices,
		SectorUtilities:             row.SectorUtilities,
		SectorIndustrials:           row.SectorIndustrials,
		SectorEnergy:                row.SectorEnergy,
		SectorHealthcare:            row.SectorHealthcare,
		EquitySector:                row.EquitySector,
		FetchedAt:                   row.FetchedAt,
	}
}

File: server-rs/src/portfolio/store/reports.rs
use std::collections::HashMap;

use anyhow::Result;
use sqlx::{Executor, Sqlite};

use crate::{database::queries, portfolio::types::AssetReport};

pub(crate) async fn upsert_report(executor: impl Executor<'_, Database = Sqlite>, report: &AssetReport) -> Result<()> {
    queries::upsert_asset_analysis_report(
        executor,
        queries::UpsertAssetAnalysisReportParams {
            asset_id: report.asset_id,
            category: &report.category,
            group_name: &report.group_name,
            cash_position: report.cash_position,
            stock_position: report.stock_position,
            bond_position: report.bond_position,
            preferred_position: report.preferred_position,
            convertible_position: report.convertible_position,
            other_position: report.other_position,
            sector_real_estate: report.sector_real_estate,
            sector_consumer_cyclical: report.sector_consumer_cyclical,
            sector_basic_materials: report.sector_basic_materials,
            sector_consumer_defensive: report.sector_consumer_defensive,
            sector_technology: report.sector_technology,
            sector_communication_services: report.sector_communication_services,
            sector_financial_services: report.sector_financial_services,
            sector_utilities: report.sector_utilities,
            sector_industrials: report.sector_industrials,
            sector_energy: report.sector_energy,
            sector_healthcare: report.sector_healthcare,
            equity_sector: report.equity_sector.as_deref(),
            fetched_at: report.fetched_at,
        },
    )
    .await
    .map_err(Into::into)
}

pub(crate) async fn reports_by_asset_ids(
    executor: impl Executor<'_, Database = Sqlite>,
    asset_ids: &[i64],
) -> Result<HashMap<i64, AssetReport>> {
    if asset_ids.is_empty() {
        return Ok(HashMap::new());
    }
    Ok(queries::asset_analysis_reports_by_asset_ids(
        executor,
        queries::AssetAnalysisReportsByAssetIDsParams { asset_ids },
    )
    .await?
    .into_iter()
    .map(|row| {
        let report = row.asset_analysis_reports;
        (report.asset_id, report)
    })
    .collect())
}

#[cfg(test)]
mod tests {
    use anyhow::Result;

    use super::{reports_by_asset_ids, upsert_report};
    use crate::{clients::yfinance::FundReport, database::dbtest, portfolio::types::AssetReport};

    #[tokio::test]
    async fn upserts_and_reads_asset_reports() -> Result<()> {
        let pool = dbtest::open().await?;
        let asset_id = sqlx::query_scalar::<_, i64>(
            "INSERT INTO assets (asset_type, identifier, classifier) VALUES ('SECURITY', 'VTI', 'PUBLIC') RETURNING id",
        )
        .fetch_one(&pool)
        .await?;
        let report = AssetReport::from((
            asset_id,
            "2026-09-06T12:00:00Z".parse()?,
            FundReport {
                category: "Large Blend".to_owned(),
                group: "US Equity".to_owned(),
                stock_position: 1.0,
                sector_technology: 0.25,
                ..Default::default()
            },
        ));

        upsert_report(&pool, &report).await?;
        let reports = reports_by_asset_ids(&pool, &[asset_id]).await?;

        assert!(reports_by_asset_ids(&pool, &[]).await?.is_empty());
        assert_eq!(reports[&asset_id].category, "Large Blend");
        assert_eq!(reports[&asset_id].sector_technology, 0.25);
        assert_eq!(reports[&asset_id].fetched_at, report.fetched_at);
        Ok(())
    }
}
